File: libs/go/identifiers/digest.go
// Copyright 2026 Mindclade. All rights reserved.
// Confidential and proprietary.

package identifiers

import (
	"bytes"
	"crypto/sha256"
	"crypto/subtle"
	"database/sql/driver"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

const (
	DigestAlgorithm  = "sha256"
	DigestBinarySize = sha256.Size
	DigestHexLength  = sha256.Size * 2
	DigestTextLength = len(DigestAlgorithm) + 1 + DigestHexLength
)

// Digest is an optional SHA-256 content digest. Its canonical text form is
// "sha256:<64 lowercase hexadecimal characters>".
//
// The unexported presence bit lets the zero value represent absence without
// making an all-zero 32-byte digest unrepresentable.
type Digest struct {
	sum     [DigestBinarySize]byte
	present bool
}

// SHA256 returns the digest of value.
func SHA256(value []byte) Digest {
	return Digest{
		sum:     sha256.Sum256(value),
		present: true,
	}
}

// SHA256String returns the digest of value's UTF-8 bytes.
func SHA256String(value string) Digest {
	return SHA256([]byte(value))
}

// SHA256Reader streams reader into a SHA-256 hash and returns the digest and
// number of bytes consumed.
func SHA256Reader(reader io.Reader) (Digest, int64, error) {
	if reader == nil {
		return Digest{}, 0, invalidValue(
			"digest reader",
			"",
			"must not be nil",
			ErrInvalidDigest,
		)
	}

	hash := sha256.New()
	count, err := io.Copy(hash, reader)
	if err != nil {
		return Digest{}, count, fmt.Errorf("identifiers: compute sha256 digest: %w", err)
	}
	digest, err := DigestFromBytes(hash.Sum(nil))
	if err != nil {
		return Digest{}, count, err
	}
	return digest, count, nil
}

// DigestFromBytes constructs a present digest from exactly 32 raw SHA-256
// bytes, including an all-zero byte sequence.
func DigestFromBytes(value []byte) (Digest, error) {
	if len(value) != DigestBinarySize {
		return Digest{}, invalidValue(
			"digest bytes",
			fmt.Sprintf("length=%d", len(value)),
			"expected exactly 32 bytes",
			ErrInvalidDigest,
		)
	}
	var digest Digest
	copy(digest.sum[:], value)
	digest.present = true
	return digest, nil
}

// ParseDigest parses canonical digest text. It rejects uppercase and alternate
// algorithm spellings to preserve one representation for signatures and keys.
func ParseDigest(value string) (Digest, error) {
	if len(value) != DigestTextLength {
		return Digest{}, invalidValue(
			"digest",
			value,
			"unexpected length",
			ErrInvalidDigest,
		)
	}
	if !strings.HasPrefix(value, DigestAlgorithm+":") {
		return Digest{}, invalidValue(
			"digest",
			value,
			"expected sha256 algorithm prefix",
			ErrInvalidDigest,
		)
	}

	hexValue := value[len(DigestAlgorithm)+1:]
	if hexValue != strings.ToLower(hexValue) {
		return Digest{}, invalidValue(
			"digest",
			value,
			"hexadecimal value must be lowercase",
			ErrInvalidDigest,
		)
	}
	decoded, err := hex.DecodeString(hexValue)
	if err != nil {
		return Digest{}, invalidValue(
			"digest",
			value,
			"contains non-hexadecimal characters",
			ErrInvalidDigest,
		)
	}
	return DigestFromBytes(decoded)
}

// MustParseDigest is ParseDigest for constants and fixtures. It panics on
// invalid input.
func MustParseDigest(value string) Digest {
	digest, err := ParseDigest(value)
	if err != nil {
		panic(err)
	}
	return digest
}

// String returns canonical digest text, or an empty string for an absent
// digest.
func (digest Digest) String() string {
	if !digest.present {
		return ""
	}
	return DigestAlgorithm + ":" + digest.Hex()
}

// Hex returns the lowercase 64-character digest value without an algorithm
// prefix, or an empty string for an absent digest.
func (digest Digest) Hex() string {
	if !digest.present {
		return ""
	}
	var output [DigestHexLength]byte
	hex.Encode(output[:], digest.sum[:])
	return string(output[:])
}

// Bytes returns an independent copy of the raw digest, or nil when absent.
func (digest Digest) Bytes() []byte {
	if !digest.present {
		return nil
	}
	output := make([]byte, DigestBinarySize)
	copy(output, digest.sum[:])
	return output
}

// IsZero reports whether digest is absent.
func (digest Digest) IsZero() bool {
	return !digest.present
}

// Valid reports whether digest is present.
func (digest Digest) Valid() bool {
	return digest.present
}

// Equal compares digest presence and bytes in constant time.
func (digest Digest) Equal(other Digest) bool {
	if digest.present != other.present {
		return false
	}
	if !digest.present {
		return true
	}
	return subtle.ConstantTimeCompare(digest.sum[:], other.sum[:]) == 1
}

// Verify reports whether value has this digest. An absent digest never
// verifies successfully.
func (digest Digest) Verify(value []byte) bool {
	return digest.present && digest.Equal(SHA256(value))
}

func (digest Digest) MarshalText() ([]byte, error) {
	if !digest.present {
		return []byte{}, nil
	}
	return []byte(digest.String()), nil
}

func (digest *Digest) UnmarshalText(value []byte) error {
	if digest == nil {
		return invalidValue("digest", string(value), "nil destination", ErrInvalidDigest)
	}
	if len(value) == 0 {
		*digest = Digest{}
		return nil
	}
	parsed, err := ParseDigest(string(value))
	if err != nil {
		return err
	}
	*digest = parsed
	return nil
}

func (digest Digest) MarshalJSON() ([]byte, error) {
	if !digest.present {
		return []byte("null"), nil
	}
	return json.Marshal(digest.String())
}

func (digest *Digest) UnmarshalJSON(value []byte) error {
	if digest == nil {
		return invalidValue("digest", string(value), "nil destination", ErrInvalidDigest)
	}
	if bytes.Equal(value, []byte("null")) {
		*digest = Digest{}
		return nil
	}

	var text string
	if err := json.Unmarshal(value, &text); err != nil {
		return invalidValue("digest", string(value), "expected a JSON string or null", ErrInvalidDigest)
	}
	return digest.UnmarshalText([]byte(text))
}

// Value implements driver.Valuer. An absent digest becomes SQL NULL.
func (digest Digest) Value() (driver.Value, error) {
	if !digest.present {
		return nil, nil
	}
	return digest.String(), nil
}

// Scan implements sql.Scanner for canonical text or raw 32-byte values.
func (digest *Digest) Scan(value any) error {
	if digest == nil {
		return invalidValue("digest", "", "nil destination", ErrInvalidDigest)
	}

	switch typed := value.(type) {
	case nil:
		*digest = Digest{}
		return nil
	case string:
		return digest.UnmarshalText([]byte(typed))
	case []byte:
		if len(typed) == DigestBinarySize {
			parsed, err := DigestFromBytes(typed)
			if err != nil {
				return err
			}
			*digest = parsed
			return nil
		}
		return digest.UnmarshalText(typed)
	default:
		return invalidValue(
			"digest",
			fmt.Sprintf("%T", value),
			"unsupported SQL source type",
			ErrInvalidDigest,
		)
	}
}

File: libs/python/identifiers/digest.py
# Copyright © 2026 Mindclade, LLC. All Rights Reserved.
# Mindclade Proprietary and Confidential.
# SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
#

"""Content digests in the one canonical form the platform addresses bytes by.

ADR-0004 makes every dataset, reference database, checkpoint, model bundle and
evidence record addressable by digest. That guarantee is only worth anything if
every producer spells a digest the same way, which before this module they did
not: nine call sites built ``"sha256:" + hashlib.sha256(...).hexdigest()`` by
hand and three mutually incompatible predicates decided whether a string was a
digest at all — a strict regex, ``startswith`` plus ``len == 71``, and
``startswith`` alone. The length form accepts uppercase hex the regex rejects, so
two validators in the same tree disagreed about the same string.

Parsing rejects uppercase hex for the reason ``libs/go/identifiers/id.go`` gives
about IDs: databases, signatures and cache keys need one byte-for-byte
representation, and accepting two spellings of one digest means a content-address
that is not actually a single address.

Unlike Go's ``identifiers.Digest``, this type carries no presence bit. Go needs
one so a zero value can mean "absent" without making an all-zero digest
unrepresentable; Python spells absence ``Digest | None``, so the type is always a
real digest and the ambiguity does not arise.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from hmac import compare_digest
from typing import IO, Final

from libs.python.errors import InvalidArgument

DIGEST_ALGORITHM: Final = "sha256"
DIGEST_BINARY_SIZE: Final = 32
DIGEST_HEX_LENGTH: Final = DIGEST_BINARY_SIZE * 2
DIGEST_PREFIX: Final = DIGEST_ALGORITHM + ":"
DIGEST_TEXT_LENGTH: Final = len(DIGEST_PREFIX) + DIGEST_HEX_LENGTH

# Streaming chunk for of_reader. Large enough that the syscall overhead disappears
# against a multi-gigabyte checkpoint, small enough not to matter for a manifest.
_READ_CHUNK: Final = 1 << 20

_LOWER_HEX: Final = frozenset("0123456789abcdef")


def is_canonical_digest(value: object) -> bool:
    """Report whether ``value`` is a canonical digest, without constructing one.

    The single predicate that replaces the three that disagreed. Checks the
    prefix, the exact length, *and* that every payload character is lowercase
    hexadecimal — the part the length-based checks omitted.
    """
    if not isinstance(value, str):
        return False
    if len(value) != DIGEST_TEXT_LENGTH or not value.startswith(DIGEST_PREFIX):
        return False
    return all(character in _LOWER_HEX for character in value[len(DIGEST_PREFIX) :])


@dataclass(frozen=True, slots=True)
class Digest:
    """A SHA-256 content digest. Canonical text form ``sha256:<64 lowercase hex>``."""

    raw: bytes

    def __post_init__(self) -> None:
        if not isinstance(self.raw, bytes):
            raise InvalidArgument(
                "digest payload must be bytes",
                reason="digest_type",
            )
        if len(self.raw) != DIGEST_BINARY_SIZE:
            raise InvalidArgument(
                f"digest must be {DIGEST_BINARY_SIZE} bytes, got {len(self.raw)}",
                reason="digest_length",
            )

    @classmethod
    def of(cls, data: bytes) -> Digest:
        """Digest ``data``."""
        if not isinstance(data, bytes):
            raise InvalidArgument("digest input must be bytes", reason="digest_input_type")
        return cls(hashlib.sha256(data).digest())

    @classmethod
    def of_text(cls, text: str) -> Digest:
        """Digest ``text``'s UTF-8 bytes.

        UTF-8 rather than the platform encoding, so the digest of a string is the
        same on every host and agrees with Go and Rust.
        """
        return cls.of(text.encode("utf-8"))

    @classmethod
    def of_reader(cls, reader: IO[bytes]) -> tuple[Digest, int]:
        """Stream ``reader`` and return its digest with the byte count consumed.

        Streamed rather than read whole: the callers that need this are hashing
        checkpoints and dataset shards, where reading the object into memory to
        hash it is the difference between working and an out-of-memory kill.
        """
        hasher = hashlib.sha256()
        consumed = 0
        while chunk := reader.read(_READ_CHUNK):
            if not isinstance(chunk, bytes):
                raise InvalidArgument(
                    "digest reader must return bytes",
                    reason="digest_reader_type",
                )
            hasher.update(chunk)
            consumed += len(chunk)
        return cls(hasher.digest()), consumed

    @classmethod
    def parse(cls, value: str) -> Digest:
        """Parse the canonical text form, rejecting anything else."""
        if not is_canonical_digest(value):
            raise InvalidArgument(
                f"digest must be {DIGEST_PREFIX}<{DIGEST_HEX_LENGTH} lowercase hex>",
                reason="digest_format",
                fields={"value": value[:128] if isinstance(value, str) else type(value).__name__},
            )
        return cls(bytes.fromhex(value[len(DIGEST_PREFIX) :]))

    @property
    def hex(self) -> str:
        """The payload without the algorithm prefix."""
        return self.raw.hex()

    @property
    def text(self) -> str:
        """The canonical text form."""
        return DIGEST_PREFIX + self.raw.hex()

    def equals(self, other: Digest) -> bool:
        """Compare in constant time.

        Digests gate artifact admission, so a comparison that returns early leaks
        how much of a forged digest was correct.
        """
        return isinstance(other, Digest) and compare_digest(self.raw, other.raw)

    def __str__(self) -> str:
        return self.text
